eos-paygd/eos-paygd.service.in

```diff
@@ -1,14 +1,22 @@
 [Unit]
 Description=Pay As You Go Daemon
 Documentation=man:eos-paygd(8)
-Before=systemd-user-sessions.service
+DefaultDependencies=no
+# Only run from initrd
+ConditionPathExists=/etc/initrd-release
+# Don't complete boot without this running
+OnFailure=emergency.target
+# Start early to reduce the chance the user is dropped to a shell before we've started
+After=basic.target
+Before=sysroot.mount
 
 [Service]
 ExecStart=@libexecdir@/eos-paygd1
-Type=dbus
-BusName=com.endlessm.Payg1
+# Type=dbus would make sense except that D-Bus isn't available in the initrd
+Type=exec
 NotifyAccess=main
-User=@DAEMON_USER@
+# We must be root to avoid systemd's post-root-pivot killing spree
+User=root
```

Comment:
It would be neat and sensible (perhaps as a later improvement) if it could drop root privileges after the killing spree has been avoided.

Reply:
Didn't know that's possible to change a process's UID. We would also have to think about the initramfs shutdown sequence, because if we're not root we won't survive the second killing spree that happens then, and presumably the service wouldn't be stopped along with the other services started in the primary root filesystem. I don't know if the way services are normally stopped is less graceful than the kill signal systemd sends them before the shutdown root pivot.

Reply:
I didn't realize we had to be root to escape the killing spree (systemd src/core/killall.c:ignore_proc makes it quite clear that this is the case, though).  dropping root privs is done through setgid/setuid - however, you can't get them back.

Only thing I can think of off-hand is moving the @ check to above the uid 0 check in a downstream systemd patch. I can't think of any reason why that's actually bad for us, but perhaps as @dsd said we should think about that for later improvement.

Reply:
I filed https://phabricator.endlessm.com/T27502

Reply:
> Only thing I can think of off-hand is moving the @ check to above the uid 0 check in a downstream systemd patch. I can't think of any reason why that's actually bad for us

We’d have to maintain that in perpetuity, and upstream behaviour might change in a different direction from our downstream patch in future. If we want to change systemd’s behaviour, it should be done upstream first. (My 2 cents; sorry for butting in.)